### PR TITLE
MongoDB snapshots

### PR DESCRIPTION
**Pull request by Tang8330**
## Changes

Implementing MongoDB snapshots to have feature parity with Debezium. 

## Checks

* [x] Test every data type
* [x] Include example
* [x] Add tests

## Data Types check 

[Comparison Gist](https://gist.github.com/Tang8330/d279bf33add5afd1245f4314edb6519c)

- [x] ObjectID
- [x] String
- [x] Int32
- [ ] Int64 (slight differences between SDKs)
- [x] Double
- [x] Decimal
- [x] SubDocument
- [x] Array
- [x] Timestamp
- [ ] Datetime (slight differences between the SDKs)
- [x] True
- [x] False
- [x] Null

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -29,7 +29,7 @@ Artie Reader reads from databases to perform historical snapshots and also reads
 | DynamoDB   | ✅        | ✅         |
 | PostgreSQL | ✅        | ❌         |
 | MySQL      | 🚧       | ❌         |
-| MongoDB    | 🚧       | ❌         |
+| MongoDB    | ✅        | ❌         |
```
Comment (nathan-artie) [inline README.md:32]: Alphabetize.

File: config/config.go [hunk 2]
```diff
@@ -107,6 +110,15 @@ func (s *Settings) Validate() error {
 		if err := s.PostgreSQL.Validate(); err != nil {
 			return fmt.Errorf("postgres validation failed: %w", err)
 		}
+
+	case SourceMongoDB:
```
Comment (nathan-artie) [inline config/config.go:114]: Move after dynamo but before postgres so they are alphabetized.

File: config/config.go [hunk 3]
```diff
@@ -107,6 +110,15 @@ func (s *Settings) Validate() error {
...
+		if s.MongoDB == nil {
+			return fmt.Errorf("mongodb config is nil")
+		}
+
+		if err := s.MongoDB.Validate(); err != nil {
+			return fmt.Errorf("mongodb validation failed: %v", err)
```
Comment (nathan-artie) [inline config/config.go:120]: `%w`

File: config/mongodb.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+import (
+	"fmt"
+	"github.com/artie-labs/reader/constants"
+	"github.com/artie-labs/transfer/lib/stringutil"
+)
+
+type MongoDB struct {
+	Host     string `yaml:"host"`
+	Username string `yaml:"userName"`
```
Comment (nathan-artie) [inline config/mongodb.go:11]: `username`

File: examples/mongodb.yaml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+source: mongodb
+mongodb:
+  host: mongodb://localhost:27017
+  userName: foo
```
Comment (nathan-artie) [inline examples/mongodb.yaml:4]: `username`

File: main.go [hunk 6]
```diff
@@ -50,6 +51,8 @@ func buildSource(cfg *config.Settings) (sources.Source, error) {
 		return dynamodb.Load(*cfg.DynamoDB)
 	case config.SourcePostgreSQL:
 		return postgres.Load(*cfg.PostgreSQL, cfg.Kafka.MaxRequestSize)
+	case config.SourceMongoDB:
```
Comment (nathan-artie) [inline main.go:54]: Alphabetize.

File: config/config.go [hunk 7]
```diff
@@ -107,6 +110,15 @@ func (s *Settings) Validate() error {
 		if err := s.PostgreSQL.Validate(); err != nil {
 			return fmt.Errorf("postgres validation failed: %w", err)
 		}
+
+	case SourceMongoDB:
+		if s.MongoDB == nil {
+			return fmt.Errorf("mongodb config is nil")
```
Comment (nathan-artie) [inline config/config.go:116]: Uppercase `MongoDB`

File: config/mongodb.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+	if c.BatchSize == 0 {
+		return constants.DefaultBatchSize
+	}
+
+	return c.BatchSize
+}
+
+func (m MongoDB) Validate() error {
+	if stringutil.Empty(m.Host, m.Database, m.Username, m.Password) {
+		return fmt.Errorf("one of the mongodb settings is empty: host, username, password, database")
```
Comment (nathan-artie) [inline config/mongodb.go:38]: Uppercase `MongoDB`.

File: sources/mongo/collection.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+		mgoMsg, err := parseMessage(result)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse message: %w", err)
+		}
+
+		mgoMsgs = append(mgoMsgs, *mgoMsg)
+	}
+
+	if err := c.cursor.Err(); err != nil {
```
Comment (nathan-artie) [inline sources/mongo/collection.go:67]: When does `c.cursor.Err()` get set? Should we be checking this inside the `for` loop?

File: sources/mongo/collection.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+		return nil, fmt.Errorf("failed to iterate over documents: %w", err)
+	}
+
+	// If the number of fetched documents is less than the batch size, we are done
+	if c.collection.GetBatchSize() > int32(len(mgoMsgs)) {
+		c.done = true
+	}
+
+	var rawMessages []lib.RawMessage
```
Comment (nathan-artie) [inline sources/mongo/collection.go:76]: We could move this inside the for loop and just build an array of raw messages.

File: sources/mongo/mongo.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
+package mongo
+
+import (
+	"context"
```
Comment (nathan-artie) [inline sources/mongo/mongo.go:4]: Reorder imports.

File: sources/mongo/mongo.go [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+func Load(cfg config.MongoDB) (*Source, error) {
+	creds := options.Credential{
+		Username: cfg.Username,
+		Password: cfg.Password,
+	}
+
+	opts := options.Client().ApplyURI(cfg.Host).SetAuth(creds).SetTLSConfig(&tls.Config{})
+	client, err := mongo.Connect(context.Background(), opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to mongodb, err: %w", err)
```
Comment (nathan-artie) [inline sources/mongo/mongo.go:31]: `MongoDB`

File: sources/mongo/mongo.go [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+	}
+
+	if err = client.Ping(context.TODO(), readpref.Primary()); err != nil {
```
Comment (nathan-artie) [inline sources/mongo/mongo.go:34]: Use the same `context.Background()` from like 29?

File: sources/mongo/mongo.go [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+		return nil, fmt.Errorf("failed to ping MongoDB, err: %w", err)
```
Comment (nathan-artie) [inline sources/mongo/mongo.go:35]: Kill `, err`.

File: sources/mongo/mongo.go [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+		db:  db,
+	}, nil
+}
+
+func (s *Source) Close() error {
+	// MongoDB doesn't need to be closed.
+	return nil
+}
+
+func (s *Source) Run(ctx context.Context, writer kafkalib.BatchWriter, statsD *mtr.Client) error {
```
Comment (nathan-artie) [inline sources/mongo/mongo.go:50]: `_` for `statsD` since it's not used.

File: README.md [hunk 16]
```diff
@@ -27,9 +27,10 @@ Artie Reader reads from databases to perform historical snapshots and also reads
 |            | Snapshot | Streaming |
 |------------|----------|-----------|
 | DynamoDB   | ✅        | ✅         |
-| PostgreSQL | ✅        | ❌         |
+| MongoDB    | ✅        | ❌         |
```
Comment (nathan-artie) [inline README.md:30]: Extra whitespace at the end of this line.

File: config/config.go [hunk 17]
```diff
@@ -62,13 +62,16 @@ type Metrics struct {
 type Source string
 
 const (
+	SourceInvalid    Source = "invalid"
 	SourceDynamo     Source = "dynamodb"
 	SourcePostgreSQL Source = "postgresql"
+	SourceMongoDB    Source = "mongodb"
```
Comment (nathan-artie) [inline config/config.go:68]: I guess we should alphabetize these too.

**Review by nathan-artie — APPROVED** [review]

File: sources/mongo/collection.go (hunk 10 above)
Comment (Tang8330) [reply sources/mongo/collection.go]: Makes the code cleaner since we have only one array too. I like it

**Review by nathan-artie — APPROVED** [review]

**Pull request merged**